File: Auto_Upload_Codeforces/Codeforces_Round_218_Div_2/B_Fox_Dividing_Cheese.cpp
#include <bits/stdc++.h>
using namespace std;

map<pair<long long,long long>,long long>cache;

int f(long long a,long long b){
    
    int val[4]={1,2,3,5};
  
    pair<long long,long long>p;
    pair<long long, long long>p2;
    p=make_pair(a,b);
    p2=make_pair(b,a);
    if (a==b){
        return 0;}
    else if (cache.find(p)!=cache.end()){
        
        return cache[p];}
    else{
        long long ans= 1000000000;
        for (int i=0;i<4;i++){
            for (int j=0;j<4;j++){
                int val1,val2;
               
                val1=val[i];
                val2=val[j];
                if (val1==1 && val2==1){
                    continue;}
                
                
                else if (a%val1==0 && b%val2==0) {
                    
                    long long inc=(val1!=1 && val2!=1) ? 2:1;
                    long long newa=a/val1;
                    long long newb=b/val2;
                    ans=min(ans,inc+f(newa,newb));
                    

                }

                }}
            cache[p]=ans;
            cache[p2]=ans;
              return ans;}
          }
int main(){
    long long a,b;
    cin>>a;
    cin>>b;
    
    long long res=f(a,b);
   
    if (res==1000000000){
        cout<<-1;}
    else{
        cout<<res;}
    return 0;
}

File: Auto_Upload_Codeforces/Codeforces_Round_218_Div_2/B_Fox_Dividing_Cheese.py
import math
import random 
from queue import Queue
import time 
cache={}
def main(a,b):
    
    val=[1,2,3,5]
    if a==b:
        return 0
    elif (a,b) in cache:
        return cache[(a,b)]

    else:
        ans=float('inf')
        for d1 in val:
            for d2 in val:
                if (d1==1 and d2==1):
                    continue 
                else:
                    val1=a/d1 
                    val2=b/d2
                    inc=2 if (d1!=1 and d2!=1) else 1 
                    if val1==int(val1) and val2==int(val2):
                        ans=min(ans,inc+main(val1,val2))
        cache[(a,b)]=ans 
        cache[(b,a)]=ans
        return ans



a,b=list(map(int,input().split()))
res=main(a,b)
if res==float('inf'):
    print(-1)
else:
    print(res)

